core/src/reservation.rs

```diff
@@ -254,6 +254,20 @@ impl GasReserver {
         }
     }
 
+    /// Check if reservation is used.
+    ///
+    /// If reservation does not exist returns `InvalidReservationId` error.
+    pub fn check_used(&mut self, id: ReservationId) -> Result<bool, ReservationError> {
+        if let Some(
+            GasReservationState::Created { used, .. } | GasReservationState::Exists { used, .. },
+        ) = self.states.get_mut(&id)
+        {
+            Ok(*used)
+        } else {
+            Err(ReservationError::InvalidReservationId)
+        }
+    }
```

Comment:
I actually expected something like this:
```suggestion
    pub fn check_not_used(&mut self, id: ReservationId) -> Result<&mut bool, ReservationError> {
        if let Some(
            GasReservationState::Created { used, .. } | GasReservationState::Exists { used, .. },
        ) = self.states.get_mut(&id)
        {
            if *used {
            	return Err(ReservationError::InvalidReservationId)
            } else {
	        Ok(used)
	    }
        } else {
            Err(ReservationError::InvalidReservationId)
        }
    }
```
So function looks exactly like original `mark_used` except it doesn't set flag.

So function in `ExtMutator` becomes:
```rust
    fn mark_reservation_used(
        &mut self,
        reservation_id: ReservationId,
    ) -> Result<(), ReservationError> {
        let _used = self.ext.context.gas_reserver.check_not_used(reservation_id)?;
	self.reservation_to_mark = Some(reservation_id);
    }
```

And function in `GasReserver` becomes:
```rust
pub fn mark_used(&mut self, id: ReservationId) -> Result<(), ReservationError> {
	let used = self.check_not_used(reservation_id)?;
	*used = true;
	Ok(())
}
```

Reply:
I suppose I implemented the suggestions too literally. Pushed the proper clean change